### ISSUE TITLE
Extend the os module

### ISSUE THREAD

**Issue by windelbouwman** [RustPython/RustPython#1175]
Currently, the python `os` module is missing many methods. Try adding a method to this module by editing the file `vm/src/stdlib/os.rs`.

This is an issue for new contributors, so please do not implement everything at once, so other people have a chance to go for it (or, actually, please implement all! >:)! ).

Note: Before working on platform-specific functions, check if they are members of `posix` or `nt`. Their members are reexported to `os`.

**Comment by mpajkowski** [comment]
Ok, I'm working on get/set uid, guid etc. stuff

**Comment by immersinn** [comment]
I'm going to take a stab at `chown` and `chroot`. 

Also, as a general question, is the current goal to match [CPython functionality here?](https://github.com/python/cpython/blob/3.8/Lib/os.py)

**Comment by coolreader18** [comment]
Because Rust's cross-platform support is so much better than C's, we decided to implement a single `_os` module rather than the separate `posix` and `nt` modules that CPython has.

**Comment by immersinn** [comment]
Ah, gotcha..so for `chown` and `chroot` methods, i'd reference the functionality and behavior in `posixmodule.c`

**Comment by coolreader18** [comment]
Yep, and you can use the `#[cfg(unix)]` attribute to make sure they're only available on unix.

**Comment by immersinn** [comment]
cool cool, thanks!

**Comment by dan-fritchman** [comment]
PR #1524 has an implementation of `os.system`.

**Comment by metaphorshear** [comment]
I was thinking of doing `wait` and `fork` if those are still available.

(Also, a question of etiquette: is it better to call dibs on things here, or just do them?)

**Comment by CircleOnCircles** [comment]
I would like to do a quick summary of what is API still missing.
1. `os.ctermid`
2. `os. environb`
3. `os.fchdir`
4. `os.fsencode`
5. `os.fsdecode`
6. `os.PathLike`
7. `os.getenvb`
8. `os.get_exec_path`
9. `os.getgrouplist`
10. `os.getgroups` *partially*
11. `os.getlogin`
12. `os.getpgrp`
13. `os.getpriority`

... 

I just scroll thought 10% on the [os docs](https://docs.python.org/3/library/os.html). There are still so many interfaces left to implement. 😛

**Comment by nicholasraphael** [comment]
Hi, new here. I was just wondering how to start working on one of these outstanding implementations without having any overlap with others? Should we open a new issue for what we want to tackle and then have it assigned to us?

Thanks!

**Comment by coolreader18** [comment]
I'm so sorry for not responding earlier @nicholasraphael, if you're still interested you can feel free to just open a PR for it; there are a lot of functions and hopefully what you choose won't conflict if anyone else is working on something. You could also just comment what you're working on in this thread.

**Comment by BolunThompson** [comment]
The [libc](https://github.com/rust-lang/libc/) rust module doesn't have an API for the [ctermid](https://man7.org/linux/man-pages/man3/ctermid.3.html) function, so that's an issue for implementing os.ctermid.

**Comment by coolreader18** [comment]
It's always possible to just do `extern "C" { fn ctermid(...) -> ...; }`; the libc crate isn't required for calling in to native functions.

**Comment by leopsidom** [comment]
I wanted to work on some of these issue. But how can I create tests for this ? Basically after I implement a method, how can I test if it's working as expected. Sorry quite new here.

**Comment by AthulMuralidhar** [comment]
hi @coolreader18 is the list  here https://github.com/RustPython/RustPython/issues/1175#issuecomment-578580002 updated? i'd like to start working on one of the API's thanks!

**Comment by coolreader18** [comment]
@AthulMuralidhar I think of that list, `ctermid fchdir getgrouplist getlogin getpriority` still need to be implemented in os.rs

**Comment by AthulMuralidhar** [comment]
thanks for the quick reply @coolreader18 i'll try to work on `ctermid` and `fchdir`

**Comment by dbrgn** [comment]
`getlogin`: #2441

**Comment by kenkoooo** [comment]
Hi, let me work on `getgrouplist`.

**Comment by kenkoooo** [comment]
Let me work on `getpriority` and `setpriority` next.

**Comment by CircleOnCircles** [comment]
I made a analysis to detemine the next important os module based on the real usage in GitHub public code. This is based on the master branch of this repo at the time of writing.

Feel free to use and modify it.

![image](https://user-images.githubusercontent.com/8089231/107870622-99580b80-6ecc-11eb-9945-c2e0547f1660.png)

[link to notebook](https://deepnote.com/project/7215d141-d8f7-4dde-b550-c7dcb65fd7c2)

**Comment by duysqubix** [comment]
Implementing os.listxattr

**Comment by hk1997** [comment]
Hi, is there something left to work on this?
Would like to pick it up.

**Comment by DimitrisJim** [comment]
Run `whats_left.sh` and take a look at the output for `os`. Pick one of the functions, add a comment here that you've picked up the function so conflicts are minimized and start working on it!

**Comment by youknowone** [comment]
I added a note to the issue body to check `posix` or `nt` for re-export.

**Comment by ecpost** [comment]
Added a few macos constants: #4110

**Comment by marvinmednick** [comment]
I'd like to take on  os.pathconf_names  - and have crated #4494  to track

**Comment by hydrogen602** [comment]
I'd like to look at getting `os.fwalk` implemented

**Comment by fanninpm** [comment]
> `whats_left.py` still shows it as missing.

If you're referring to what's on the website, the website gets updated every Friday (at least, in my timezone).

**Comment by youknowone** [comment]
Thank you everyone to participate in this marathon for os module.

os module function are almost done. A few of them are left, but having an issue for that is not worth as before.
